contrib/epee/include/net/abstract_tcp_server2.inl

```diff
@@ -1222,6 +1297,75 @@ POP_WARNINGS
 
     _dbg3("Connected success to " << adr << ':' << port);
 
+    const epee::net_utils::ssl_support_t ssl_support = new_connection_l->get_ssl_support();
+    if (ssl_support == epee::net_utils::ssl_support_t::e_ssl_support_enabled || ssl_support == epee::net_utils::ssl_support_t::e_ssl_support_autodetect)
+    {
+      // Handshake
+      MDEBUG("Handshaking SSL...");
+      if (!new_connection_l->handshake(boost::asio::ssl::stream_base::client))
+      {
+        if (ssl_support == epee::net_utils::ssl_support_t::e_ssl_support_autodetect)
+        {
+          boost::system::error_code ignored_ec;
+          sock_.shutdown(boost::asio::ip::tcp::socket::shutdown_both, ignored_ec);
+          sock_.close();
+          ssl = false;
+          return true;
```

Comment:
You need both (3 cases can happen): the physical connection can succeed, if the peer does not support SSL, the logical connection fails. Then you know you should retry without SSL, which is that case: return true because the connection succeeds, but set ssl to true to say "but we actually need SSL". I suppose you'd use a boost::optional for that :)

Reply:
Yes, forgot about the case of connection failure with SSL enabled (differed from connection success with SSL failure). In this instance, I would probably use an enum.

Reply:
Well, I might have advocated for a `boost::optional<bool>` in the past, but the difference between null, true, and false seems kind of funky here.

Reply:
I'll make it an enum then.

Reply:
done